lib/lightstep/propagation/lightstep_propagator.rb

```diff
@@ -0,0 +1,127 @@
+#frozen_string_literal: true
+
+module LightStep
+  module Propagation
+    class LightStepPropagator
+      CARRIER_TRACER_STATE_PREFIX = 'ot-tracer-'
+      CARRIER_SPAN_ID = 'ot-tracer-spanid'
+      CARRIER_TRACE_ID = 'ot-tracer-traceid'
+      CARRIER_SAMPLED = 'ot-tracer-sampled'
+      CARRIER_BAGGAGE_PREFIX = 'ot-baggage-'
+
+      # Inject a SpanContext into the given carrier
+      #
+      # @param spancontext [SpanContext]
+      # @param format [OpenTracing::FORMAT_TEXT_MAP, OpenTracing::FORMAT_BINARY]
+      # @param carrier [Carrier] A carrier object of the type dictated by the specified `format`
+      def inject(span_context, format, carrier)
+        case format
+        when OpenTracing::FORMAT_TEXT_MAP
+          inject_to_text_map(span_context, carrier)
+        when OpenTracing::FORMAT_BINARY
+          warn 'Binary inject format not yet implemented'
+        when OpenTracing::FORMAT_RACK
+          inject_to_rack(span_context, carrier)
+        else
+          warn 'Unknown inject format'
+        end
+      end
+
+      # Extract a SpanContext from a carrier
+      # @param format [OpenTracing::FORMAT_TEXT_MAP, OpenTracing::FORMAT_BINARY, OpenTracing::FORMAT_RACK]
+      # @param carrier [Carrier] A carrier object of the type dictated by the specified `format`
+      # @return [SpanContext] the extracted SpanContext or nil if none could be found
+      def extract(format, carrier)
+        case format
+        when OpenTracing::FORMAT_TEXT_MAP
+          extract_from_text_map(carrier)
+        when OpenTracing::FORMAT_BINARY
+          warn 'Binary join format not yet implemented'
+          nil
+        when OpenTracing::FORMAT_RACK
+          extract_from_rack(carrier)
+        else
+          warn 'Unknown join format'
+          nil
+        end
+      end
+
+      private
+
+      def inject_to_text_map(span_context, carrier)
+        if trace_id = trace_id_from_ctx(span_context)
+          carrier[self.class::CARRIER_TRACE_ID] = trace_id
+        end
+        carrier[self.class::CARRIER_SPAN_ID] = span_context.id
+        carrier[self.class::CARRIER_SAMPLED] = sampled_flag_from_ctx(span_context)
+
+        span_context.baggage.each do |key, value|
+          carrier[self.class::CARRIER_BAGGAGE_PREFIX + key] = value
+        end
+      end
+
+      def extract_from_text_map(carrier)
+        # If the carrier does not have both the span_id and trace_id key
+        # skip the processing and just return a normal span
+        if !carrier.has_key?(self.class::CARRIER_SPAN_ID) || !carrier.has_key?(self.class::CARRIER_TRACE_ID)
```

Comment:
For this and baggage, is `has_key`/`start_with` sensitive to upper/lowercase? I've seen before where the header keys will come in sometimes uppercase, sometimes lowercase.

Reply:
`extract_from_text_map` is using case-sensitive comparisons. This is not new, and has been the original behavior. 

All HTTP header parsing will go through `extract_from_rack`. Rack is the webserver interface for Ruby and it normalizes headers for you. If a header comes in as `x-b3-spanid` it will be normalized to `HTTP_X_B3_SPANID`. So, for the http header case (aka FORMAT_RACK) keys will be processed in a case-insensitive way.

Reply:
ah, rad. thanks for clearing it up!